tests/std/tests/P3349R1_contiguous_iterators_to_pointers/test.cpp

```diff
@@ -0,0 +1,463 @@
+    } else {
+        try {
+            func();
+            assert(false && "should throw, but returned");
+        } catch (const safe_iter_out_of_bounds_err&) {
+        }
+    }
+}
+
+#define PASS(expr) single_test_pass<nothrow>([&]() { (void) (expr); })
+#define FAIL(expr) single_test_fail<nothrow>([&]() { (void) (expr); })
+
+#define EXPR_WRAP_IS(expr)       ([&](auto&& i, auto&& s) { (void) (expr); })
+#define EXPR_WRAP_IN(expr)       ([&](auto&& i, auto&& n) { (void) (expr); })
+#define EXPR_WRAP_ISD(expr)      ([&](auto&& i, auto&& s, auto&& dst) { (void) (expr); })
+#define EXPR_WRAP_IS_DI_DS(expr) ([&](auto&& i, auto&& s, auto&& dst_i, auto&& dst_s) { (void) (expr); })
+
+template <class T, size_t range_size, bool nothrow>
+void test() {
+    using iter       = safe_iter<T, nothrow>;
+    using const_iter = safe_iter<const T, nothrow>;
+    static_assert(contiguous_iterator<iter>);
+    static_assert(contiguous_iterator<const_iter>);
+
+    // Allow unsafe sentinels
+    static_assert(sized_sentinel_for<typename iter::pointer, iter>);
+    static_assert(sized_sentinel_for<typename const_iter::pointer, const_iter>);
+
+    // Allows trivial container reuse for uninitialized memory algorithms
+    static_assert(is_trivially_destructible_v<T>);
+
+    vector<T> container_read(range_size);
+    iota(container_read.begin(), container_read.end(), static_cast<T>(0));
+
+    vector<T> container_read_unique(range_size);
+    iota(container_read_unique.begin(), container_read_unique.end(), static_cast<T>(range_size));
+    assert(ranges::find_first_of(container_read, container_read_unique) == ranges::end(container_read));
+
+    auto container_write   = container_read;
+    auto container_write_2 = container_read;
+
+#define GEN_RANGE_ITER(prefix, suffix, type, range, iter_type)                      \
+    [[maybe_unused]] const auto [prefix##_first_##suffix, prefix##_last_##suffix] = \
+        iter_type::get_iters(range, range_##type);                                  \
+    [[maybe_unused]] const auto prefix##_dist_##suffix =                            \
+        ranges::distance(prefix##_first_##suffix, prefix##_last_##suffix);
+#define GEN_RANGE_ITERS(prefix, range, iter_type)       \
+    GEN_RANGE_ITER(prefix, s, small, range, iter_type)  \
+    GEN_RANGE_ITER(prefix, m, medium, range, iter_type) \
+    GEN_RANGE_ITER(prefix, l, large, range, iter_type)
+
+    // Iterators with the same suffix refer to ranges of the same size
+    GEN_RANGE_ITERS(r, container_read, const_iter)
+    GEN_RANGE_ITERS(r2, container_read_unique, const_iter)
+    GEN_RANGE_ITERS(w, container_write, iter)
+    GEN_RANGE_ITERS(w2, container_write_2, iter)
+
+    assert(r_first_s < r_first_m && r_first_m < r_first_l && r_last_s < r_last_m && r_last_m < r_last_l);
+
+#undef GEN_RANGE_ITERS
+#undef GEN_RANGE_ITER
+
+    static constexpr auto tests_base = [](auto&& func, auto&& first, auto&& last_or_dist) {

```

Comment:
Non-arrays can be plain `constexpr`, as the `static` doesn't make a difference. (For arrays it makes a considerable difference.)